packages/flutter/lib/src/material/material.dart

```diff
@@ -585,6 +587,13 @@ class _InkFeatures extends SingleChildRenderObjectWidget {
                 ..absorbHitTest = absorbHitTest;
     assert(vsync == renderObject.vsync);
   }
+
+  @override
+  void didUnmountRenderObject(_RenderInkFeatures renderObject) {
+    while (renderObject._inkFeatures?.isNotEmpty ?? false) {
+      renderObject._inkFeatures!.first.dispose();
+    }
```

Comment:
this looks like it leaves `renderObject._inkFeatures` with a bunch of pointers to disposed objects, maybe instead we should call a method on the render object to clean itself up, and have that method do something like:

```dart
for (InkFeature feature in _inkFeatures)
  feature.dispose();
_inkFeatures = null;
```

Reply:
I don't think so, because InkFeature has _RenderInkFeatures, which calls _RenderInkFeatures._removeFeature() when call InkFeature.dispose().

```dart
  void _removeFeature(InkFeature feature) {
    assert(_inkFeatures != null);
    _inkFeatures!.remove(feature);
    markNeedsPaint();
  }
```


I improved this part of the code and added comments to explain it.

Reply:
Or do you like this?

```dart
    if (_inkFeatures != null) {
      for (int i = _inkFeatures!.length - 1; i >= 0; i--){
        _inkFeatures![i].dispose();
      }
    }
```

I really like the use of `list?.isNotEmpty ? false`, it looks very intuitive and clean.